Source/BoundaryConditions/WarpX_PEC.H

```diff
@@ -0,0 +1,166 @@
+     *        field value of the guard cells outside the domain boundary are set equal
+     *        and opposite to the field in the valid cells at their mirrored locations.
+     *        The number or depth of guard cells updated is equal to the shape factor of
+     *        particles in each dimension.
+     *
+     * \param[in] icomp        component of the Efield being updated
+                               (0=x, 1=y, 2=z in 3D) and (0=x, 1=z in 2D)
+     * \param[in] dom_lo       index value of the lower domain boundary (cell-centered)
+     * \param[in] dom_hi       index value of the higher domain boundary (cell-centered)
+     * \param[in] ijk_vec      indices along the x(i), y(j), z(k) of Efield Array4
+     * \param[in] shape_factor particle shape factor in each dimension
+     * \param[in] Efield       field data to be updated if (ijk) is at the boundary or a guard cell
+     * \param[in] stag         staggering of the field data being updated.
+     * \param[in] fbndry_lo    Field boundary type at the lower boundaries
+     * \param[in] fbndry_hi    Field boundary type at the upper boundaries
+     */
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    void SetTangentialEfield (const int icomp, const amrex::IntVect & dom_lo,
+                                const amrex::IntVect &dom_hi,
+                                const amrex::IntVect &ijk_vec,
+                                const amrex::IntVect &shape_factor,
+                                amrex::Array4<amrex::Real> const& Efield,
+                                const amrex::IntVect& stag,
+                                amrex::GpuArray<int, 3> const& fbndry_lo,
+                                amrex::GpuArray<int, 3> const& fbndry_hi )
+    {
+        // Fields in guard cells set equal and opposite to valid cells
+        amrex::IntVect ijk_guard = ijk_vec;
+        amrex::IntVect ijk_valid = ijk_vec;
+        for (int idim = 1; idim < AMREX_SPACEDIM; ++idim) {
+            // index of dimension
+            const int dir = ( ( (icomp+idim)<AMREX_SPACEDIM)? (icomp+idim) : ( (icomp+idim) - AMREX_SPACEDIM));
+            // loop over sides , iside = 0 (lo), iside = 1 (hi)
+            for (int iside = 0; iside < 2; ++iside) {
+                const int ibndry = dir + iside*AMREX_SPACEDIM;
+                const int bndry_cell = ( (ibndry < AMREX_SPACEDIM) ?
+                    (ijk_vec[dir] == dom_lo[dir]) : (ijk_vec[dir] == dom_hi[dir] + stag[dir]));
+                const int is_tangent_to_PEC =
+                    ( ( ibndry < AMREX_SPACEDIM) ? is_boundary_PEC(fbndry_lo, dir)
+                                                 : is_boundary_PEC(fbndry_hi, dir) );
+                // if boundary is PEC, guard cells must be equal and opposite to valid cells
+                if ( bndry_cell == 1 && is_tangent_to_PEC==1) {
+                    for (int is = 0; is < shape_factor[dir]; ++is) {
```

Comment:
This routine is being called from inside a ParallelFor loop and so already can loop over the guard cells. Could this be restructured so that this routine only sets a single element and lets the ParallelFor loop over the guard cells?

Reply:
Yes @dpgrote . 
I initially thought it would complicate the logic and code because I would need to add code to
1. find if the ijk is ON the boundary to set or outside
2. How many cells is it outside by, so that we get the valid cell field accordingly.
So I did it this way.

But, I could change add in additional logic here if it is confusing with the ParallelFor

Reply:
Fixed this in this commit 
4d7ed59